libfs/fs.go

```diff
@@ -0,0 +1,273 @@
+			return nil, errors.Errorf("%s is not a directory",
+				path.Join(parts[:i]...))
+		}
+	}
+
+	log := config.MakeLogger("")
+	log.CDebugf(ctx, "Made new FS for TLF=%s, subdir=%s",
+		tlfHandle.GetCanonicalName(), subdir)
+
+	return &FS{
+		ctx:      ctx,
+		config:   config,
+		root:     n,
+		log:      log,
+		deferLog: log.CloneWithAddedDepth(1),
+	}, nil
+}
+
+func (fs *FS) lookupOrCreateFile(
+	dir libkbfs.Node, filename string, flag int, perm os.FileMode) (
+	libkbfs.Node, libkbfs.EntryInfo, error) {
+	n, ei, err := fs.config.KBFSOps().Lookup(fs.ctx, dir, filename)
+	switch errors.Cause(err).(type) {
+	case libkbfs.NoSuchNameError:
+		// The file doesn't exist yet; create if requested
+		if flag&os.O_CREATE == 0 {
+			return nil, libkbfs.EntryInfo{}, err
+		}
+		fs.log.CDebugf(
+			fs.ctx, "Creating %s since it doesn't exist yet", filename)
+		excl := libkbfs.NoExcl
+		if flag&os.O_EXCL != 0 {
+			excl = libkbfs.WithExcl
+		}
+		isExec := (perm & 0100) != 0
+		n, ei, err = fs.config.KBFSOps().CreateFile(
+			fs.ctx, dir, filename, isExec, excl)
+		switch errors.Cause(err).(type) {
+		case libkbfs.NameExistsError:
+			// Someone made it already; recurse to try the lookup again.
+			fs.log.CDebugf(
+				fs.ctx, "Attempting lookup again after failed create")
+			return fs.lookupOrCreateFile(dir, filename, flag, perm)
+		case nil:
+			return n, ei, nil
+		default:
+			return nil, libkbfs.EntryInfo{}, err
+		}
+	case nil:
+		// If we were supposed to have exclusively-created this file,
+		// we must fail.
+		if flag&os.O_CREATE != 0 && flag&os.O_EXCL != 0 {
+			return nil, libkbfs.EntryInfo{},
+				errors.New("Exclusive create failed because the file exists")
+		}
+
+		// Make sure this is a file.
+		if !ei.Type.IsFile() {
+			return nil, libkbfs.EntryInfo{},
+				errors.Errorf("%s is not a file", filename)
+		}
+		return n, ei, nil
+	default:
+		return nil, libkbfs.EntryInfo{}, err
+	}
+}
+
+// OpenFile implements the billy.Filesystem interface for FS.
+func (fs *FS) OpenFile(filename string, flag int, perm os.FileMode) (
+	f billy.File, err error) {
+	fs.log.CDebugf(
+		fs.ctx, "OpenFile %s, flag=%d, perm=%o", filename, flag, perm)
+	defer func() {
+		fs.deferLog.CDebugf(fs.ctx, "OpenFile done: %+v", err)
+	}()
+
+	parts := strings.Split(filename, "/")
+	n := fs.root
+	for i := 0; i < len(parts)-1; i++ {
+		p := parts[i]
+		var ei libkbfs.EntryInfo
+		n, ei, err = fs.config.KBFSOps().Lookup(fs.ctx, n, p)
+		if err != nil {
+			return nil, err
+		}
+		if ei.Type != libkbfs.Dir {
+			return nil, errors.Errorf("%s is not a directory",
+				path.Join(parts[:i]...))
+		}
+	}
+
+	fName := parts[len(parts)-1]
+	n, ei, err := fs.lookupOrCreateFile(n, fName, flag, perm)
+	if err != nil {
+		return nil, err
+	}
+
+	if flag&os.O_TRUNC != 0 {
+		err := fs.config.KBFSOps().Truncate(fs.ctx, n, 0)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	offset := int64(0)
+	if flag&os.O_APPEND != 0 {
+		// TODO: worry about overflow.
```

Comment:
TODO: wow! That's a big offset!

Kidding aside, this seems worth it to do right away rather than a TODO:
```go
if ei.Size >= uint64(1<<63) {
  return nil, errors.New("offset too large")
}
```